### PR TITLE
Polars: alias

### PR DESCRIPTION
**Pull request by Shoeboxam**

### CODE REVIEW

**Comment by Shoeboxam** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opendp/opendp/1591?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1592** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1592?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1591** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1591?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1534** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1534?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1539** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1539?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1540** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1540?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1586** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1586?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1590** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1590?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1584** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1584?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1542** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1542?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1498** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1498?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1479** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1479?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1496](https://github.com/opendp/opendp/pull/1496) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1496?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#908** <a href="https://app.graphite.dev/github/pr/opendp/opendp/908?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#913](https://github.com/opendp/opendp/pull/913) <a href="https://app.graphite.dev/github/pr/opendp/opendp/913?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#890** <a href="https://app.graphite.dev/github/pr/opendp/opendp/890?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#856](https://github.com/opendp/opendp/pull/856) <a href="https://app.graphite.dev/github/pr/opendp/opendp/856?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#829** <a href="https://app.graphite.dev/github/pr/opendp/opendp/829?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#854](https://github.com/opendp/opendp/pull/854) <a href="https://app.graphite.dev/github/pr/opendp/opendp/854?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#888](https://github.com/opendp/opendp/pull/888) <a href="https://app.graphite.dev/github/pr/opendp/opendp/888?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1519** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1519?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1506](https://github.com/opendp/opendp/pull/1506) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1506?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1566** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1566?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: rust/src/transformations/make_stable_expr/expr_alias/mod.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+                    if col == old_name {
+                        name.to_string()
+                    } else {
+                        col
+                    }
+                })
+                .collect();
+            (new_k.len() == old_len).then_some((new_k, v))
+        })
+        .collect();
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_alias/mod.rs:61]: A little more explanation here: A comment at the top, or pull out the internals in a named function. I'm not really sure what's happening with the `len`.

File: rust/src/transformations/make_stable_expr/expr_alias/test.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+use crate::{
+    domains::{AtomDomain, LazyFrameDomain, SeriesDomain},
+    metrics::SymmetricDistance,
+};
+
+use super::*;
+
+#[test]
+fn test_behavior() -> Fallible<()> {
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_alias/test.rs:12]: Unless I'm missing something, this isn't a test of our code, it's just making sure Polars works as expected? Maybe explain the motivation... are we not confident about the behavior of polars? Or is it a reminder of all the ways aliases could be used?

If it's worth keeping, maybe reorganize these two tests, so that each unit test will just do one operation with Polars, and then the same thing with our extension. Then, if we do realize there's an edge case we forgot, the addition will be one new unit that checks Polars out-of-the-box, and then checks our code.

Sorry if I'm reading this wrong.

Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/expr_alias/test.rs:12]: Yes, the aliasing transformation needs to change the output domain in the same way that polars changes dataframes when aliasing.

File: rust/src/domains/polars/frame/mod.rs [hunk 3]
```diff
@@ -210,6 +210,10 @@ impl<F: Frame> FrameDomain<F> {
         self.margins.insert(grouping_keys, margin);
         Ok(self)
     }
+
+    pub fn schema(&self) -> Schema {
```
Comment (mccalluc) [inline rust/src/domains/polars/frame/mod.rs:214]: I think this is only used for tests? Could imagine it being defined for a narrower scope.

File: rust/src/transformations/make_stable_expr/expr_alias/mod.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+) -> Fallible<Transformation<ExprDomain, ExprDomain, M, M>>
+where
+    M::InnerMetric: DatasetMetric,
+    M::Distance: Clone,
+    (ExprDomain, M): MetricSpace,
+    Expr: StableExpr<M, M>,
+{
+    let Expr::Alias(input, name) = expr else {
+        return fallible!(MakeTransformation, "expected alias expression");
+    };
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_alias/mod.rs:35]: This comes up in most of these, but I don't really understand why the type couldn't be tighter

**Review by mccalluc — DISMISSED** [review]

File: rust/src/domains/polars/frame/mod.rs (hunk 3 above)
Comment (Shoeboxam) [reply rust/src/domains/polars/frame/mod.rs:214]: It will also be used in the context PR.

File: rust/src/transformations/make_stable_expr/expr_alias/mod.rs (hunk 4 above)
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/expr_alias/mod.rs:35]: The type can only as specific as `Expr`, not a specific variant like `Expr::Alias`.
